File: lib/domain/cultivation_plan/gateways/cultivation_plan_rest_plan_read_gateway.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Gateways
      # REST 計画 read（表ごと）。複合は Interactor + domain mapper。
      class CultivationPlanRestPlanReadGateway
        # @return [Dtos::CultivationPlanRestPlanHeaderSnapshot]
        def find_plan_header_snapshot_by_plan_id(plan_id:)
          raise NotImplementedError, "#{self.class}##{__method__}"
        end

        # @return [Array<Dtos::CultivationPlanRestPlanFieldRowSnapshot>]
        def list_rest_plan_field_row_snapshots_by_plan_id(plan_id:)
          raise NotImplementedError, "#{self.class}##{__method__}"
        end

        # @return [Array<Dtos::CultivationPlanRestPlanCropRowSnapshot>]
        def list_rest_plan_crop_row_snapshots_by_plan_id(plan_id:)
          raise NotImplementedError, "#{self.class}##{__method__}"
        end

        # @return [Array<Dtos::CultivationPlanRestPlanCultivationRowSnapshot>]
        def list_rest_plan_cultivation_row_snapshots_by_plan_id(plan_id:)
          raise NotImplementedError, "#{self.class}##{__method__}"
        end

        # @return [Array<Integer>]
        def list_palette_crop_ids_by_plan_id(plan_id:)
          raise NotImplementedError, "#{self.class}##{__method__}"
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/gateways/cultivation_plan_rest_plan_read_gateway.rs
//! Ruby: `Domain::CultivationPlan::Gateways::CultivationPlanRestPlanReadGateway`

use crate::cultivation_plan::dtos::rest_plan_snapshots::{
    CultivationPlanRestPlanCropRowSnapshot, CultivationPlanRestPlanCultivationRowSnapshot,
    CultivationPlanRestPlanFieldRowSnapshot, CultivationPlanRestPlanHeaderSnapshot,
};

pub trait CultivationPlanRestPlanReadGateway: Send + Sync {
    fn find_plan_header_snapshot_by_plan_id(
        &self,
        plan_id: i64,
    ) -> Result<CultivationPlanRestPlanHeaderSnapshot, Box<dyn std::error::Error + Send + Sync>>;

    fn list_rest_plan_field_row_snapshots_by_plan_id(
        &self,
        plan_id: i64,
    ) -> Result<Vec<CultivationPlanRestPlanFieldRowSnapshot>, Box<dyn std::error::Error + Send + Sync>>;

    fn list_rest_plan_crop_row_snapshots_by_plan_id(
        &self,
        plan_id: i64,
    ) -> Result<Vec<CultivationPlanRestPlanCropRowSnapshot>, Box<dyn std::error::Error + Send + Sync>>;

    fn list_rest_plan_cultivation_row_snapshots_by_plan_id(
        &self,
        plan_id: i64,
    ) -> Result<Vec<CultivationPlanRestPlanCultivationRowSnapshot>, Box<dyn std::error::Error + Send + Sync>>;

    fn list_palette_crop_ids_by_plan_id(
        &self,
        plan_id: i64,
    ) -> Result<Vec<i64>, Box<dyn std::error::Error + Send + Sync>>;
}
